### PR TITLE
1.12 Branch

### PR DESCRIPTION
**Pull request by OfficialTomPG**
1.12 build! Changed various files to support new Forge mechanics in 1.12.
As per canitpz comment on original PR (to wrong branch), I will take a look at backporting some of the file changes for consistency, as well as looking at bugs.

### CODE REVIEW

**Review by canitzp — APPROVED** [review]

**Comment by canitzp** [comment]
I want the opinion of @lorddusk and @way2muchnoise too, before I merge it

File: src/main/java/hardcorequesting/team/PlayerEntry.java [hunk 1]
```diff
@@ -118,7 +119,6 @@ public void write(JsonWriter out) throws IOException {
         out.endObject();
     }
 
-    @SideOnly(Side.SERVER)
```
Comment (way2muchnoise) [inline src/main/java/hardcorequesting/team/PlayerEntry.java]: You can't just remove this, this line of code doesn't work client side (on MP when connected to a server)

**Review by way2muchnoise — COMMENTED** [review]
No json recipes? Also seems like you messed up the Russian and Chinese translation.

File: src/main/java/hardcorequesting/team/PlayerEntry.java (hunk 1 above)
Comment (ghost) [reply src/main/java/hardcorequesting/team/PlayerEntry.java]: 100% agree but the client continued to crash with this line in place, stating the method was missing. Mainly when trying to pick up entity items.

Also, I've found json recipes tend to be temperamental and not always work, so used the built in GameRegistry recipe registration.

Feel free to reject the Russina/Chinese translation files - just need to update the names of the items (couldn't get rendering working when they were metadata items)

Comment (way2muchnoise) [reply src/main/java/hardcorequesting/team/PlayerEntry.java]: Then you need to fix the code used on item pickup and not remove an annotation that is put there for that exact reason. Now that pickup code will only work in SP and will still crash in MP.

**Comment by Fundryi** [comment]
So 1.12 & 1.12.1 is coming soon? :)

**Review by lorddusk — COMMENTED** [review]
I'm fine with the port if way is.

**Comment by OfficialTomPG** [comment]
commit 8bbe6faf26993b7de7287f405f19b6298f1d3f78 tweaks support for PickupEvent - works in both SP and MP (tested) - Team creation UI is broken and not working as intended

**Comment by ryantheleach** [comment]
@ArclightTW What's currently wrong with team creation UI?

**Comment by OfficialTomPG** [comment]
Clicking 'Create Team' doesn't seem to do anything and you can't get into the Team screen to invite other players/accept invites/etc.

Feels like a sync issue with the server but I'm not an expert by any stretch of the imagination so couldn't point to a specific cause with 100% certainty.

**Comment by canitzp** [comment]
So do you can fix the create Team bug or is your port finished at this state, so I can merge it?

**Comment by ryantheleach** [comment]
@ArclightTW I'm a complete beginner to (forge) modding, but I've had some experience with Sponge / Sponge plugins.

I've got some time I can put towards trying to find the issue / testing tonight.

**Comment by OfficialTomPG** [comment]
@ryantheleach Thanks, I'll see if I can trace the code back too to work out why Team creation isn't working properly. From my tests, it's the only thing not working at this stage (though I haven't run a server-side test yet either).

**Comment by OfficialTomPG** [comment]
@lorddusk @way2muchnoise @canitzp Party(Team) creation UI is now working and I have tested the build on a local server, too, and it withstands general usage, including parties with multiple members and quest completions for all.

If you're happy to merge this, I believe everything is stable.

**Comment by ryantheleach** [comment]
I'll try to get around to testing quest creation and saving tonight.

**Comment by OfficialTomPG** [comment]
I'm happy with this being merged now and a 1.12 build being made available if everyone else is.

**Comment by ryantheleach** [comment]
I didn't get around to testing, but it's probably fine to merge considering some of the open issues that were open previously. Worst comes to worse you might need to make some maintenance updates.

**Comment by canitzp** [comment]
I'll test it this weekend and if it works without crashing, I would ask if it is good for a public release

**Comment by lorddusk** [comment]
If you consider it fit for a public release, I'll allow it :P



Met vriendelijke groeten,

Tim Langen

**Comment by canitzp** [comment]
@ArclightTW can you try to fix a bug, that if you create a questline at the client you can't simply copy the config/hqm folder to the server and launch it, cause then no quest is loaded?

**Comment by OfficialTomPG** [comment]
@canitzp I just tested this is my dev environment and could copy/past config/hqm from the client config to server config folder and all sets/quests were copied across?

**Comment by canitzp** [comment]
Have you changed some config options? I used 'runClient' created some quests and then restarted the client and made a server with 'runServer' and then no quests gets loaded into the game

**Comment by OfficialTomPG** [comment]
Nothing changed, just copied the folders over in config. I’ll take some screenshots tomorrow to see if we can narrow down the cause, since it’s definitely a bug if it’s ownly sometimes occurring.

**Comment by ryantheleach** [comment]
I'd suggest trying in a non-dev environment.

**Comment by OfficialTomPG** [comment]
Ran this again in a non-dev environment running just HQM;
1. Started client with HQM installed
2. Created simple quest with task, bag and reputation
3. Started Forge server
4. Copied HQM mod to server along with all contents of config/hqm
5. Connected to server, quest/bag/reputation working as intended.

N.B. I did note that HQM spits out a warning about quests being incompatible when running on the server, but didn't notice any changes when playing on the server v.s. the client.

Not sure I can replicate the problem you are having @canitzp?

**Comment by OfficialTomPG** [comment]
It's been noted to me that building on 1.12.2 (like I've been doing) makes the mod incompatible with 1.12.1 and 1.12; are you running on an earlier MC version @canitzp?

I'll get a 1.12 workspace set up, rebuild and see if I run into the same problem and see if I can fix if I do.

**Comment by canitzp** [comment]
I have simply cloned your repo and after setupDecompWorkspace I used the runClient and runServer commands andI also haven't to move the config files from client to server, cause if I run them with their corresponding run commands they use the same config directory, maybe this can cause the issue too

**Comment by canitzp** [comment]
I haven't tested it again, but I'll accept it now and build a version for 1.12 release. Do you still want to help developing HQM with things like bug fixing @ArclightTW?

**Comment by OfficialTomPG** [comment]
@canitzp Yeah, I’m happy to help out with bug fixing for HQM :)

**Comment by canitzp** [comment]
@ArclightTW it would be nice if you are on the HQM discord: https://discord.gg/P89tzu

**Pull request merged**